In [ ]:
Contents
1. Import libraries
2. Import data
3. Check for mixed-type data in your dfords dataframe.
4. Run a check for missing values in dfords dataframe.
5. Run a check for duplicate values in your dfords data.

01. Importing libraries

In [1]:
#Import libraries
import pandas as pd
import numpy as np
import os

02. Importing data

In [3]:
df_prods = pd.read_csv(r'C:\Users\kelly\OneDrive\Desktop\Original Data\products.csv', index_col = False)

In [4]:
dfords = pd.read_csv(r'C:\Users\kelly\OneDrive\Desktop\Prepared Data\orders_wrangled.csv', index_col = False)

In [5]:
#Create a dataframe
df_test = pd.DataFrame()

In [6]:
#Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [7]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [8]:
#Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [9]:
df_test['mix'] = df_test['mix'].astype('str')

In [10]:
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [11]:
df_nan = df_prods[df_prods['product_name'].isnull() ==True]

In [12]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [13]:
df_prods.shape

(49693, 5)

In [14]:
df_prods_clean = df_prods[df_prods['product_name'].isnull()==False]

In [15]:
df_prods_clean.shape

(49677, 5)

In [16]:
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [17]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [18]:
df_prods_clean.shape

(49677, 5)

In [19]:
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [21]:
df_prods_clean_no_dups.shape

(49672, 5)

In [24]:
#2 Run the df.describe() function on your df_prods dataframe.
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

I think the prices column needs further investigation.  The min value is 1.  The max value is 99999.  The mean is 9.99.  The 50% values is 7.10.  Something seems very off about the max price.  There must be some outliers somewhere.

In [31]:
dfords.head()

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0           0   2539329  2539329    prior             1          2   
1           1   2398795  2398795    prior             2          3   
2           2    473747   473747    prior             3          3   
3           3   2254736  2254736    prior             4          4   
4           4    431534   431534    prior             5          4   

   order_hour_of_day  days_since_prior_order  
0                  8                     NaN  
1                  7                    15.0  
2                 12                    21.0  
3                  7                    29.0  
4                 15                    28.0

In [32]:
#3 Check for mixed-type data in your dfords dataframe.
for col in dfords.columns.tolist():
  weird = (dfords[[col]].applymap(type) != dfords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (dfords[weird]) > 0:
    print (col)

#4. I didn't find any mixed-type data.

In [34]:
#5 Run a check for missing values in dfords dataframe.
dfords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

There are 206,209 missing values in the days_since_prior_order column.  This could be because there are brand new customers without any prior orders.

In [35]:
df_nan = dfords[dfords['days_since_prior_order'].isnull()==True]

In [36]:
df_nan

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0                 0   2539329  2539329    prior             1          2   
11               11   2168274  2168274    prior             1          2   
26               26   1374495  1374495    prior             1          1   
39               39   3343014  3343014    prior             1          6   
45               45   2717275  2717275    prior             1          3   
...             ...       ...      ...      ...           ...        ...   
3420930     3420930    969311   969311    prior             1          4   
3420934     3420934   3189322  3189322    prior             1          3   
3421002     3421002   2166133  2166133    prior             1          6   
3421019     3421019   2227043  2227043    prior             1          1   
3421069     3421069   3154581  3154581    prior             1          3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 8 columns]

In [39]:
dfords.shape

(3421083, 8)

In [40]:
#6 Address the missing values using an apporopriate method.
dfords_clean = dfords[dfords['days_since_prior_order'].isnull()==False]

In [41]:
dfords_clean.shape

(3214874, 8)

I chose to filter out missing data by creating a new dataframe.  I'm not sure it is safe to drop the missing rows and would rather be safe than sorry later.

In [42]:
#7 Run a check for duplicate values in your dfords data.
df_dups = dfords_clean[dfords_clean.duplicated()]

In [43]:
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

I'm confused by this output.  Does this mean there are no duplicates when it doesn't show any information, just the headers?

In [44]:
dfords_clean.shape

(3214874, 8)

In [45]:
dfords_clean_no_dups = dfords_clean.drop_duplicates()

In [46]:
dfords_clean_no_dups.shape

(3214874, 8)

#8 Address the duplicates using an appopriate method.
I didn't get any results for duplicates when I ran the above command.  I went ahead and created a new dataframe that would have dropped any duplicates.  I ran the shape command before and after trying to drop the duplicates and the results are the same.  This shows that were not any duplicates in this dataframe.

In [47]:
dfords_clean_no_dups.to_csv(r'C:\Users\kelly\OneDrive\Desktop\Prepared Data/orders_checked.csv')